packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/Converters.kt

```diff
@@ -217,77 +318,73 @@ internal fun <T : BaseRealmObject> realmObjectConverter(
     realmReference: RealmReference
 ): RealmValueConverter<T> {
     return object : PassThroughPublicConverter<T>() {
-        override fun fromRealmValue(realmValue: RealmValue): T? =
-            // TODO OPTIMIZE We could lookup the companion and keep a reference to
-            //  `companion.newInstance` method to avoid repeated mediator lookups in Link.toRealmObject()
+        // TODO OPTIMIZE We could lookup the companion and keep a reference to
+        //  `companion.newInstance` method to avoid repeated mediator lookups in Link.toRealmObject()
+        override fun fromRealmValue(realmValue: RealmValue?): T? =
             realmValueToRealmObject(realmValue, clazz, mediator, realmReference)
 
-        override fun toRealmValue(value: T?): RealmValue =
-            realmObjectToRealmValueWithImport(value as BaseRealmObject?, mediator, realmReference)
+        override fun MemTrackingAllocator.toRealmValue(value: T?): RealmValue = when (value) {
+            null -> transportOf()
+            else -> transportOf(realmObjectToRealmReferenceOrError(value) as RealmObjectInterop)
+        }
     }
 }
 
 internal inline fun <T : BaseRealmObject> realmValueToRealmObject(
-    realmValue: RealmValue,
+    realmValue: RealmValue?,
     clazz: KClass<T>,
     mediator: Mediator,
     realmReference: RealmReference
 ): T? {
-    return realmValue.value?.let {
-        (it as Link).toRealmObject(
+    return realmValue?.getLink()
+        ?.toRealmObject(
             clazz,
             mediator,
             realmReference
         )
-    }
-}
-
-// Will return a RealmValue wrapping a managed realm object reference (or null) or throw when
-// called with an unmanaged object
-internal inline fun realmObjectToRealmValueOrError(
-    value: BaseRealmObject?,
-): RealmValue {
-    return RealmValue(
-        value?.let {
-            value.runIfManaged { this }
-                ?: throw IllegalArgumentException("Cannot lookup unmanaged objects in realm")
-        }
-    )
 }
 
-// Will return a RealmValue wrapping a managed realm object reference (or null). If the object
-// is unmanaged it will be imported according to the update policy. If the object is an outdated
-// object it will will throw an error.
-internal inline fun realmObjectToRealmValueWithImport(
+// Will return a managed realm object reference or null. If the object is unmanaged it will be
+// imported according to the update policy. If the object is an outdated object it will throw an
+// error.
+internal inline fun realmObjectToRealmReference(
```

Comment:
I think we should keep the `WithImport` in the name, so that it is explicit that we import objects through this method.